### PR TITLE
feat: add new toml parser

### PR DESCRIPTION
**Pull request by amrbashir**
## Motivation

Toml parsers in JS are lossy and don't preserve comments, spaces and relative order or items (see #194).

## Approach

Leveraging Rust and WASM, we can create a small library that uses `toml_edit` crate which has an excellent support for the shortcomings of those JS equivalents.

The new library exports a `TomlDocument` class that can parse toml and then can be used to access arbitray keys through `tomlDocument.field.anotherField` or `tomlDocument['field']['anotherField']` or `tomlDocument.get('field.anotherField')` and a value can be set using `tomlDocument.field.anotherField = newValue` or `tomlDocument['field']['anotherField'] = newValue` or `tomlDocument.set('field.anotherField', newValue)`

### TODOs and Open Questions

- [x] Integrate with the codebase

### CODE REVIEW

File: packages/toml/index.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    set(t, innerProp, newValue) {
+      return target.set(prop + "." + innerProp, newValue);
+    },
+  });
+}
+
+class TomlDocument {
+  inner;
+
+  constructor(toml) {
```
Comment (jbolda) [inline packages/toml/index.js:18]: Would it be worth adding a `.parse()` method that calls the constructor and sets things up? It seems the `new` keyword is a bit out of vogue these days.

Comment (amrbashir) [reply packages/toml/index.js:18]: yeah sure

File: packages/toml/package.json [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    "pkg/covector_toml.js",
+    "pkg/covector_toml.d.ts",
+    "dist/index.js",
+    "dist/index.d.ts"
+  ],
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "build": "wasm-pack build --release --no-pack --target nodejs && rimraf pkg/.gitignore",
+    "postbuild": "cp ./pkg/covector_toml_bg.wasm ../action/dist && cp ./pkg/covector_toml.js ../action/dist && npm run build:ts",
```
Comment (jbolda) [inline packages/toml/package.json:15]: Does the `action` `pkg` command not end up handling this? I would have expected it to move it to the `dist` for us.

File: packages/toml/package.json [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    "pkg/covector_toml_bg.wasm",
+    "pkg/covector_toml.js",
+    "pkg/covector_toml.d.ts",
+    "dist/index.js",
+    "dist/index.d.ts"
+  ],
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
+  "scripts": {
+    "build": "wasm-pack build --release --no-pack --target nodejs && rimraf pkg/.gitignore",
```
Comment (jbolda) [inline packages/toml/package.json:14]: I think I would make this a `prebuild`, and then change `build:ts` to `build`. I think this means that we will get a `wasm-pack` prior to other builds without having to change any of our workflows.

File: packages/toml/package.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    "postbuild": "cp ./pkg/covector_toml_bg.wasm ../action/dist && cp ./pkg/covector_toml.js ../action/dist && npm run build:ts",
+    "build:debug": "wasm-pack build --dev --no-pack --target nodejs && rimraf pkg/.gitignore && npm run postbuild",
+    "build:ts": "tsc -b",
+    "clean": "rimraf dist tsconfig.tsbuildinfo node_modules",
+    "pretest": "npm run build",
```
Comment (jbolda) [inline packages/toml/package.json:19]: We do run a `build` in CI before running the tests, so I am wondering if we could possibly remove this. However, would we want this for first run locally with tests?

File: packages/toml/pkg/covector_toml_bg.wasm.d.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+/* tslint:disable */
+/* eslint-disable */
+export const memory: WebAssembly.Memory;
```
Comment (jbolda) [inline packages/toml/pkg/covector_toml_bg.wasm.d.ts:3]: I think we want to `.gitignore` this `pkg` directory, and allow it to be committed in the `action`. We don't _need_ it here, ultimately only in the `action` directory.

File: packages/toml/package.json (hunk 3 above)
Comment (jbolda) [reply packages/toml/package.json:14]: Ahh, forgot we aren't doing a workspace build at the root. I think on top of this, we also need to add `"prebuild": "npm run prebuild -w @covector/toml",` at the root.

File: packages/toml/package.json (hunk 2 above)
Comment (amrbashir) [reply packages/toml/package.json:15]: it doesn't unfortunately

File: packages/toml/package.json (hunk 4 above)
Comment (amrbashir) [reply packages/toml/package.json:19]: I guess we could remove and just run it locally if needed

**Comment by jbolda** [comment]
The status is failing because this PR is opened through a fork, and the token used does not support the appropriate permissions. This would be solved through a bot token, but as it is opt-in, we won't solve for it wholesale at the current time.

**Review by jbolda — APPROVED** [review]

**Pull request merged**